### PR TITLE
FHAC-585:Synchronize Audit Proxy cxf client calls for cxf webservice client call and fix Java Impl Audit proxy client call

### PR DESCRIPTION
**Pull request by alameluchidambaram**
1. Synchronize cxf webservice Audit proxy client calls.
2. Fix Java Impl Audit proxy call and initialize sessionfactory from the servlets.
3. Add Bean TransactionManagementType for EJB calls.

### CODE REVIEW

File: Product/Production/Gateway/CONNECTGatewayWeb/src/main/java/gov/hhs/fha/nhinc/gateway/servlet/InitServlet.java [hunk 1]
```diff
@@ -63,8 +62,10 @@ public void init(ServletConfig config) throws ServletException {
         LOG.debug("InitServlet start...");
         executor = Executors.newFixedThreadPool(ExecutorServiceHelper.getInstance().getExecutorPoolSize());
         largeJobExecutor = Executors.newFixedThreadPool(ExecutorServiceHelper.getInstance()
-                .getLargeJobExecutorPoolSize());
-
+            .getLargeJobExecutorPoolSize());
+        //Initialize HibernateUtil when CONNECTGatewayWeb is initialized required for AuditRepo JavaImpl EJB calls.
+        //Do not Remove this ...
+        SessionFactory session = HibernateUtil.getSessionFactory();
```
Comment (christophermay07) [inline Product/Production/Gateway/CONNECTGatewayWeb/src/main/java/gov/hhs/fha/nhinc/gateway/servlet/InitServlet.java]: Do we need the local variable session anywhere?  Otherwise, just call getSessionFactory without assigning it to anything.

File: Product/Production/Services/AuditRepositoryCore/src/test/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAOTest.java [hunk 2]
```diff
@@ -24,8 +24,10 @@
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-package gov.hhs.fha.nhinc.hibernate;
+package gov.hhs.fha.nhinc.auditrepository.hibernate;
 
+import gov.hhs.fha.nhinc.auditrepository.hibernate.AuditRepositoryRecord;
+import gov.hhs.fha.nhinc.auditrepository.hibernate.AuditRepositoryDAO;
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/test/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAOTest.java]: There are two added imports, but no lines added in this PR that use them.  Can they be removed?

**Comment by christophermay07** [comment]
Great PR, only a couple minor comments.  Also: tomorrow, can you explain to the team why the EJB needs to be bean-managed instead of container-managed?  Seems like a good learning opportunity for all of us.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/util/HibernateUtil.java [hunk 3]
```diff
@@ -43,12 +45,13 @@
...
+
     static {
         try {
             // Create the SessionFactory from hibernate.cfg.xml
 
             sessionFactory = new Configuration().configure(getConfigFile()).buildSessionFactory();
-        } catch (Throwable ex) {
+        } catch (HibernateException ex) {
             // Make sure you log the exception, as it might be swallowed
             LOG.error("Initial SessionFactory creation failed." + ex);
```
Comment (TabassumJafri) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/util/HibernateUtil.java]: Can we log ex.getLocalizedMessage() here

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyWebServiceSecuredImpl.java [hunk 4]
```diff
@@ -50,13 +50,14 @@
...
+    private AcknowledgementType result = new AcknowledgementType();
+    private LogEventSecureRequestType secureRequest = new LogEventSecureRequestType();
 
+    @Override
     public AcknowledgementType auditLog(LogEventRequestType request, AssertionType assertion) {
         LOG.debug("Entering AuditRepositoryProxyWebServiceSecured.auditLog(...)");
-        AcknowledgementType result = new AcknowledgementType();
 
-        LogEventSecureRequestType secureRequest = new LogEventSecureRequestType();
         if (request.getAuditMessage() == null) {
```
Comment (TabassumJafri) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyWebServiceSecuredImpl.java:65]: Need to handle the scenario when request.getAuditMessage() == null what value should be populated for LogEventSecureRequestType .setAuditMessage().  
LogEventSecureRequestType is a jaxb class and has AuditMessage as a required field.

**Comment by alameluchidambaram** [comment]
@christophermay07 
The Container managed transactions works for cxf webservice client call where those classes initiate DAO call are managed by container. The JavaImpl call does not works since the DAO call is a pojo call  and therefore Bean managed transactions works for Java Impl Audit proxy calls.

**Comment by TabassumJafri** [comment]
Good with PR with no further comments

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java [hunk 5]
```diff
@@ -145,13 +147,13 @@ public List queryAuditRepositoryOnCriteria(String eUserId, String ePatientId, Da
...
-                        new Date(endDate.getTime())));
+                    new Date(endDate.getTime())));
             } else if (startDate != null && endDate == null) {
                 aCriteria.add(Expression.ge("timeStamp", new Date(startDate.getTime())));
             }
             queryList = aCriteria.list();
-        } catch (Exception e) {
-            LOG.error("Exception in AuditLog.get() occured due to :" + e.getMessage());
+        } catch (HibernateException e) {
+            LOG.error("Exception in AuditLog.get() occured due to :" + e.getLocalizedMessage(), e);
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/hibernate/AuditRepositoryDAO.java]: If you need to make another commit on this PR, can you fix this typo?  Should be "occurred".

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyJavaImpl.java [hunk 6]
```diff
@@ -38,7 +38,9 @@
  */
 public class AuditRepositoryProxyJavaImpl implements AuditRepositoryProxy {
 
-	protected AuditRepositoryOrchImpl getAuditRepositoryOrchImpl() {
+    private LogEventSecureRequestType securedRequest = new LogEventSecureRequestType();
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyJavaImpl.java]: Can this be final, or can it be reassigned at some point?

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyWebServiceSecuredImpl.java [hunk 7]
```diff
@@ -50,44 +50,45 @@
 
     private static final Logger LOG = Logger.getLogger(AuditRepositoryProxyWebServiceSecuredImpl.class);
 
-    private WebServiceProxyHelper oProxyHelper = new WebServiceProxyHelper();
+    private final WebServiceProxyHelper oProxyHelper = new WebServiceProxyHelper();
+    private AcknowledgementType result = new AcknowledgementType();
+    private LogEventSecureRequestType secureRequest = new LogEventSecureRequestType();
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyWebServiceSecuredImpl.java]: Same question as above, can these be final?

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyWebServiceUnsecuredImpl.java [hunk 8]
```diff
@@ -45,42 +44,48 @@
...
     private static final Logger LOG = Logger.getLogger(AuditRepositoryProxyWebServiceUnsecuredImpl.class);
 
-    private WebServiceProxyHelper oProxyHelper = new WebServiceProxyHelper();
+    private final WebServiceProxyHelper oProxyHelper = new WebServiceProxyHelper();
+    private AcknowledgementType result = new AcknowledgementType();
 
+    @Override
     public AcknowledgementType auditLog(LogEventRequestType request, AssertionType assertion) {
+
         LOG.debug("Entering AuditRepositoryProxyWebServiceUnsecuredImpl.auditLog(...)");
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyWebServiceUnsecuredImpl.java]: Can this (and other instances of "Entering Class.Method") either be changed to trace level, or removed in favor of a more meaningful log message?

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyWebServiceUnsecuredImpl.java [hunk 9]
```diff
@@ -45,42 +44,48 @@
...
-                    result = (AcknowledgementType) client.invokePort(AuditRepositoryManagerPortType.class, "logEvent", request);
+                CONNECTClient<AuditRepositoryManagerPortType> client = CONNECTCXFClientFactory.getInstance()
+                    .getCONNECTClientUnsecured(portDescriptor, url, assertion);
 
-                } else {
-                    LOG.error("Failed to call the web service (" + NhincConstants.AUDIT_REPO_SERVICE_NAME
-                            + ").  The URL is null.");
+                synchronized (result) {
+                    result = (AcknowledgementType) client.invokePort(AuditRepositoryManagerPortType.class,
+                        "logEvent", request);
```
Comment (christophermay07) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyWebServiceUnsecuredImpl.java]: "logEvent" should be a constant.

File: Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyJavaImpl.java (hunk 6 above)
Comment (alameluchidambaram) [inline Product/Production/Services/AuditRepositoryCore/src/main/java/gov/hhs/fha/nhinc/auditrepository/nhinc/proxy/AuditRepositoryProxyJavaImpl.java]: The LogEventSecureRequestType values are set later.

**Comment by christophermay07** [comment]
I've reviewed this PR, endorsing for merge without further comment.
